backoff: support reset backoff period

wu0407 opened the pull request:
reset backoff period when call Wait after a long time, default BackoffResetDuration is 10 minute.

I think this change does not affect calling the `wait` method

Fixes: #21936

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

Please ensure your pull request adheres to the following guidelines:

- [x] For first time contributors, read [Submitting a pull request](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#submitting-a-pull-request)
- [x] All code is covered by unit and/or runtime tests where feasible.
- [x] All commits contain a well written commit description including a title,
      description and a `Fixes: #XXX` line if the commit addresses a particular
      GitHub issue.
- [x] If your commit description contains a `Fixes: <commit-id>` tag, then
      please add the commit author[s] as reviewer[s] to this issue.
- [x] All commits are signed off. See the section [Developer’s Certificate of Origin](https://docs.cilium.io/en/stable/contributing/development/contributing_guide/#dev-coo)
- [x] Provide a title or release-note blurb suitable for the release notes.
- [x] Thanks for contributing!

<!-- Description of change -->

```release-note
support reset backoff period
```

gandro commented on pkg/backoff/backoff.go:
```diff
@@ -114,6 +119,16 @@ func ClusterSizeDependantInterval(baseInterval time.Duration, numNodes int) time
 func (b *Exponential) Wait(ctx context.Context) error {
+	resetDuration := b.BackoffResetDuration
+	if resetDuration == time.Duration(0) {
+		resetDuration = 10 * time.Minute
```
I think this is incorrect if `b.Max` is larger than 10 minutes, it will reset the attempts before `b.Max` is even reached. 

One option would be to set this to `2 * b.Max`, but I think it might be better to not assume a default value and have users set `BackoffResetDuration` explicitly.

wu0407 replied:
> One option would be to set this to `2 * b.Max`, but I think it might be better to not assume a default value and have users set `BackoffResetDuration` explicitly.

I think only do reset when BackoffResetDuration is set and if BackoffResetDuration less than b.Max  force set BackoffResetDuration to 2 * b.Max

gandro replied:
Yeah, let's only do the reset if the value is set. Let's also just ignore it if it is less than Max, since that is a user error. I added some sample documentation in the comment above.

gandro reviewed (commented):
Thanks. However, I do think to fix #21936 it would be better to explicitly reset the backoff counter, i.e. add a `Reset()` method to `Exponential` and have the backoff reset once `MaintainIPPool` succeeds.

gandro commented on pkg/backoff/backoff.go:
```diff
@@ -56,6 +56,11 @@ type Exponential struct {
 	Name string
 
+	// BackoffResetDuration is since lastBackoffStart exceeded the duration time that reset the backoff period
+	BackoffResetDuration time.Duration
```
Let's just name this `BackoffReset` (to be consistent with Min/Max, which are also not named MinDuration or MaxDuration) and document it in terms of how it is used from a user's perspective. For example:

```suggestion
	// BackoffReset will reset the exponential back-off if no attempt is made for the amount of time specified here. 
	// Needs to be larger than the Max duration, otherwise it will be ignored to avoid accidental resets.
	// If unspecified, no reset is performed.
	BackoffReset time.Duration
```

gandro reviewed (commented):
Maybe let's do both, i.e. add ResetBackoff and a manual `Reset()`?

I agree that having the attempt counter reset after a healthy duration is easier and might even be preferable to a manual reset. But we should have the user be able to perform manual reset if needed as well.

In the end, I think the `Exponential` wasn't originally designed for re-use. We should check if there are other uses besides IPAM that reuse the same struct and fix it accordingly.

bimmlerd commented on pkg/backoff/backoff.go:
```diff
@@ -112,8 +119,23 @@ func ClusterSizeDependantInterval(baseInterval time.Duration, numNodes int) time
 // Wait waits for the required time using an exponential backoff
 func (b *Exponential) Wait(ctx context.Context) error {
+	resetDuration := b.BackoffReset
+	if resetDuration != time.Duration(0) && resetDuration > b.Max {
```
it seems that these nested if statements are all just ANDed, is there a reason to not write them as a single if statement?

bimmlerd commented on pkg/backoff/backoff.go:
```diff
@@ -56,6 +56,13 @@ type Exponential struct {
+	// BackoffReset will reset the exponential back-off if no attempt is made for the amount of time specified here.
+	// Needs to be larger than the Max duration, otherwise it will be ignored to avoid accidental resets.
+	// If unspecified, no reset is performed.
+	BackoffReset time.Duration
```
Maybe call it `ResetAfter`? I don't think the `Backoff` prefix adds a lot, given that this is a backoff struct, and just `Reset` is mildly unclear and collides with the method.

bimmlerd commented on pkg/ipam/node_manager.go:
```diff
@@ -311,10 +311,11 @@ func (n *NodeManager) Update(resource *v2.CiliumNode) (nodeSynced bool) {
+			BackoffReset: 10 * time.Minute,
 		}
 		poolMaintainer, err := trigger.NewTrigger(trigger.Parameters{
 			Name:            fmt.Sprintf("ipam-pool-maintainer-%s", resource.Name),
```
(cannot comment on the relevant line 327 since it's not changed)
Should we be resetting the backoff manually here too, given that we suceeded in maintaining the pool?

wu0407 replied:
I expect backoff reset in the last wait to this execution exceeds the set reset threshold. not reset only successfully in maintaining the pool

bimmlerd replied:
I'm not sure I understand your reply correctly. I interpret it as follows: you expect the backoff to be reset once we `Wait` but the `ResetAfter` time period has expired, not whenever we manage to maintain the pool successfully.

If that's the intended behaviour, fair enough- but why is it the intended behaviour? Do we expect to only fail intermittently in maintaining the pool, which would lead to the backoff being reset too quickly? To be clear, I'm not advocating to remove the ResetAfter, just wondering why we would not reset manually once we've succeeded.

wu0407 replied:
>I'm not sure I understand your reply correctly. I interpret it as follows: you expect the backoff to be reset once we Wait but the ResetAfter time period has expired, not whenever we manage to maintain the pool successfully.

Yes, all right.

>If that's the intended behaviour, fair enough- but why is it the intended behaviour? Do we expect to only fail intermittently in maintaining the pool, which would lead to the backoff being reset too quickly? To be clear, I'm not advocating to remove the ResetAfter, just wondering why we would not reset manually once we've succeeded.

If cloud API is unstable, such as rate limit access cloud api,  it will cause sometimes success and sometimes failure and may be cause an infinite loop in this situation.

bimmlerd replied:
> If cloud API is unstable, such as rate limit access cloud api, it will cause sometimes success and sometimes failure and may be cause an infinite loop in this situation.

I'm not sure I follow what you mean by an infinite loop: the trigger is executed repeatedly anyway, right?

But yes, if we are being rate limited and start hammering the API whenever we have a single success, I can see how that is not ideal. The difference between manual reset and reset because of ResetAfter is just the amount of time we wait before hammering again though right? So in effect we're trying to make the cloud rate limit happier by waiting longer before we start going full throttle again. Maybe it'd be smarter to _decrease_ the backoff on success, instead of resetting it completely?

wu0407 replied:
>I'm not sure I follow what you mean by an infinite loop: the trigger is executed repeatedly anyway, right?

If maintpool success (assign ip or release ip  or create interface) or failed that trigger manager resync. It will re-trigger maintIPPool.

https://github.com/cilium/cilium/blob/f59df8580b559aa122c3f0ff9d322360dbff7ef9/pkg/ipam/node.go#L946-L957

wu0407 replied:
>Maybe it'd be smarter to decrease the backoff on success, instead of resetting it completely?

yes, it is smart but more complicated

bimmlerd replied:
all good points, lgtm.

bimmlerd reviewed (commented):
generally LGTM, modulo nits.
I was initially not convinced that adding a self-healing mechanism is necessary, but seeing that forgetting to reset the backoff is an easy mistake to make, I think it's a good idea.

wu0407 committed:
backoff: support reset backoff period

reset backoff period when the last wait to this execution exceeds the reset threshold

Fixes: #21936

Signed-off-by: xiaoqing <xiaoqingnb@gmail.com>

jrajahalme reviewed (approved):
LGTM
